unrouted_handler.go

```
@@ -0,0 +1,626 @@
+
+		// Do not continue with this element if no key and value or presented
+		if len(parts) != 2 {
+			continue
+		}
+
+		// Ignore corrent element if the value is no valid base64
+		key := parts[0]
+		value, err := base64.StdEncoding.DecodeString(parts[1])
+		if err != nil {
+			continue
+		}
+
+		meta[key] = string(value)
+	}
+
+	return meta
+}
+
+// Serialize a map of strings into the Upload-Metadata header format used in the
+// response for HEAD requests.
+// e.g. Upload-Metadata: name bHVucmpzLnBuZw==,type aW1hZ2UvcG5n
+func serializeMeta(meta map[string]string) string {
+	header := ""
+	for key, value := range meta {
+		valueBase64 := base64.StdEncoding.EncodeToString([]byte(value))
+		header += key + " " + valueBase64 + ","
+	}
+
+	// Remove trailing comma
+	if len(header) > 0 {
+		header = header[:len(header)-1]
+	}
+
+	return header
+}
+
+// Parse the Upload-Concat header, e.g.
+// Upload-Concat: partial
+// Upload-Concat: final; http://tus.io/files/a /files/b/
+func parseConcat(header string) (isPartial bool, isFinal bool, partialUploads []string, err error) {
+	if len(header) == 0 {
+		return
+	}
+
+	if header == "partial" {
+		isPartial = true
+		return
+	}
+
+	l := len("final; ")
+	if strings.HasPrefix(header, "final; ") && len(header) > l {
+		isFinal = true
+
+		list := strings.Split(header[l:], " ")
+		for _, value := range list {
+			value := strings.TrimSpace(value)
+			if value == "" {
+				continue
+			}
+
+			id := extractIDFromPath(value)
+			if id == "" {
+				err = ErrInvalidConcat
+				return
+			}
+
+			partialUploads = append(partialUploads, id)
+		}
+	}
+
+	// If no valid partial upload ids are extracted this is not a final upload.
+	if len(partialUploads) == 0 {
+		isFinal = false
+		err = ErrInvalidConcat
+	}
+
+	return
+}
+
+// extractIDFromPath pulls the last segment from the url provided
+func extractIDFromPath(url string) string {
+	result := reExtractFileID.FindStringSubmatch(url)
+	if len(result) != 2 {
+		return ""
```

This should probably throw an error, shouldn't it?